src/lazy/binary/immutable_buffer.rs

```diff
@@ -157,6 +159,257 @@ impl<'a> ImmutableBuffer<'a> {
+    /// to be correctly interpreted as positive, negative, or beyond the bounds of the 64 bit
+    /// limitation.  
+    fn read_flex_primitive_as_uint(
+        self,
+        label: &'static str,
+        support_sign_extension: bool,
+    ) -> ParseResult<'a, FlexUInt> {
+        // A closure that generates an incomplete data result at the current offset. This can be invoked
+        // in a variety of early-return cases in this method.
+        let incomplete = || IonResult::incomplete(label, self.offset);
+
+        let bytes_available = self.bytes().len();
+        if bytes_available == 0 {
+            return incomplete();
+        }
+
+        // The `from_le_bytes` method we use to interpret data requires at least 8 bytes to be available.
+        // There can be 1-2 bytes of header for a u64, leading to a maximum size of 10 bytes. If the input
+        // buffer doesn't have at least 10 bytes, copy its contents into a temporary buffer that's
+        // padded with 0 bytes. We round the size of the temp buffer to 16 as it produces slightly
+        // nicer assembly than 10.
+        let mut buffer = [0u8; 16];
+        let bytes = if bytes_available >= 10 {
+            self.bytes()
+        } else {
+            buffer[0..bytes_available].copy_from_slice(self.bytes());
+            &buffer[..]
+        };
+
+        let first_byte = bytes[0];
+        // If the first byte is not zero, the FlexUInt is 7 or fewer bytes.
+        if first_byte != 0 {
+            let num_encoded_bytes = first_byte.trailing_zeros() as usize + 1;
+            // Note that `bytes_available` is the number of bytes in the original unpadded input.
+            // Our buffer may be 16 bytes long but only `bytes_available` of those are meaningful.
+            if bytes_available < num_encoded_bytes {
+                return incomplete();
+            }
+            // At this point, we know the original input contained all of the FlexUInt's bytes.
+            // We can call `read_small_flex_uint` with the now-padded version of the buffer.
+            // It will discard any bytes that are not part of the FlexUInt.
+            let flex_uint = Self::read_small_flex_uint(bytes);
+            let remaining = self.consume(flex_uint.size_in_bytes());
+            return Ok((flex_uint, remaining));
+        }
+
+        // If we reach this point, the first byte was a zero. The FlexUInt is at least 9 bytes in size.
+        // We need to inspect the second byte to see how many more prefix bits there are.
+        if bytes_available < 2 {
+            return incomplete();
+        }
+        let second_byte = bytes[1];
+
+        if second_byte & 0b11 == 0b00 {
+            // The flag bits in the second byte indicate at least two more bytes, meaning the total
+            // length is more than 10 bytes. We're not equipped to handle this.
+            return IonResult::decoding_error("found a >10 byte VarUInt too large to fit in a u64");
```

Comment:
```suggestion
            return IonResult::decoding_error("found a >10 byte FlexUInt too large to fit in a u64");
```
...although this will still be an inaccurate message if you try to read a `FlexInt` (which delegates to this implementation) of more than 10 bytes.